src/Elastic.Apm/Report/PayloadSenderV2.cs

```diff
@@ -0,0 +1,147 @@
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Elastic.Apm.Api;
+using Elastic.Apm.Config;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Model.Payload;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace Elastic.Apm.Report
+{
+	/// <summary>
+	/// Responsible for sending the data to the server. Implements Intake V2.
+	/// Each instance creates its own thread to do the work. Therefore instances should be reused if possible.
+	/// </summary>
+	public class PayloadSenderV2 : IPayloadSender
+	{
+		private readonly ScopedLogger _logger;
+		private readonly Service _service;
+
+		private readonly JsonSerializerSettings _settings;
+
+		private static readonly int DnsTimeout = (int)TimeSpan.FromMinutes(1).TotalMilliseconds;
+
+		private readonly BatchBlock<object> _eventQueue =
+			new BatchBlock<object>(1, new GroupingDataflowBlockOptions
+				{ BoundedCapacity = 1_000_000 });
+
+		private readonly HttpClient _httpClient;
+
+
+		public void QueueTransaction(ITransaction transaction)
+		{
+			_eventQueue.Post(transaction);
+			_eventQueue.TriggerBatch();
+		}
+
+		public void QueueSpan(ISpan span) => _eventQueue.Post(span);
+
+		public void QueueError(IError error) => _eventQueue.Post(error);
+
+		public PayloadSenderV2(IApmLogger logger, IConfigurationReader configurationReader, Service service, HttpMessageHandler handler = null)
+		{
+			_settings = new JsonSerializerSettings { ContractResolver = new CamelCasePropertyNamesContractResolver(), Formatting = Formatting.None };
+			_service = service;
+			_logger = logger?.Scoped(nameof(PayloadSenderV2));
+
+			var serverUrlBase = configurationReader.ServerUrls.First();
+			var servicePoint = ServicePointManager.FindServicePoint(serverUrlBase);
+
+			servicePoint.ConnectionLeaseTimeout = DnsTimeout;
+			servicePoint.ConnectionLimit = 20;
+
+			_httpClient = new HttpClient(handler ?? new HttpClientHandler())
+			{
+				BaseAddress = serverUrlBase
+			};
+
+			if (configurationReader.SecretToken != null)
+			{
+				_httpClient.DefaultRequestHeaders.Authorization =
+					new AuthenticationHeaderValue("Bearer", configurationReader.SecretToken);
+			}
+
+			Task.Factory.StartNew(
+				async () =>
+				{
+					while (true)
+					{
+						try
+						{
+							await DoWork();
+						}
+						catch (TaskCanceledException ex)
+						{
+							_logger.LogDebugException(ex);
+						}
+					}
+				}, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+		}
+
+		private async Task DoWork()
+		{
+			while (true)
+			{
+				var queueItems = await _eventQueue.ReceiveAsync();
+				try
+				{
+					var metadata = new Metadata { Service = _service };
+					var metadataJson = JsonConvert.SerializeObject(metadata, _settings);
+					var json = "{\"metadata\": " + metadataJson + "}" + "\n";
+
+					foreach (var item in queueItems)
+					{
+						var serialized = JsonConvert.SerializeObject(item, _settings);
+						switch (item)
```

Comment:
Hi @gregkalapos ,
Maybe it`s better to use JObject ? 
I mean smth like : 
```
            var x = JObject.FromObject(obj, new JsonSerializer()
            {
                ContractResolver = new CamelCasePropertyNamesContractResolver(),
                Formatting = Formatting.None
            });
            var jObj = new JObject()
            {
                {"metedata",x}
            };```

Reply:
That code looks good, but the `new JObject()` part is an additional allocation which we don't have if we just append to a `StringBuilder`. I know that my code is less elegant, but the JSON we create in this case is very minimal, so I think it's worth to avoid the allocation for this.